agents/superAgent.py

```diff
@@ -22,22 +25,43 @@ def run(self):
         """
         Function that runs the agents.
         """
+        strategist = [agent for agent in threading.enumerate()
+                      if agent.name == 'Strategist'][0]
+        if strategist:
+            self.input_queue = strategist.input_queue
+            self.output_queue = strategist.output_queue
+        else:
+            print('Agents play without the strategist.')
+
         while True:
             # Receive a message.
             msg = self.receive_msg()
+
             time.sleep(0.6)  # uncomment to add delay per step in agent.
 
             if msg:
                 # Parse the response.
                 if msg["type"] == "request-action":
-                    self.beliefs.update(msg, self._user_id)
+                    # Get the request id
+                    request_id = self._get_request_id(msg)
+                    agent_id = self._user_id
+                    # Update beliefs
+                    self.beliefs.update(msg, agent_id)
+
+                    # Send a message to the strategist that if has updated its
+                    # beliefs. Agent can continue when all agents updated their
+                    # belief.
+                    if hasattr(self, 'input_queue'):
+                        self.input_queue.put(('update', self))
+                        self.input_queue.join()
+
+                        self.input_queue.put(('merge', self))
```

Comment:
Maybe you could make a function in Strategist with a clear 'update_beliefs' kind of name, as it may now be unclear that the agent is updating it's beliefs.

Reply:
Beliefs get updated in the SuperAgent. This is simply to message the strategist that the beliefs have been updated